common/state/sebio.cpp

```diff
@@ -0,0 +1,204 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "sebio.h"
+#include "crypto.h"
+#include "state.h"
+#include "types.h"
+#include "crypto.h"
+#include "c11_support.h"
+#include "error.h"
+
+#include "packages/block_store/block_store.h"
+
+#ifdef DEBUG
+    #define SAFE_LOG(LEVEL, FMT, ...) Log(LEVEL, FMT, ##__VA_ARGS__)
+#else // DEBUG not defined
+    #define SAFE_LOG(LEVEL, FMT, ...)
+#endif // DEBUG
+
+state_status_t sebio_fetch_from_block_store(
+    uint8_t* block_id,
+    size_t block_id_size,
+    sebio_crypto_algo_e crypto_algo,
+    uint8_t** block,
+    size_t* block_size);
+state_status_t sebio_evict_to_block_store(
+    uint8_t* block,
+    size_t block_size,
+    sebio_crypto_algo_e crypto_algo,
+    ByteArray& idOnEviction);
+
+
+//########### internal sebio (Secure Block IO) context ####################
+/*
+    The sebio context describes operations to be performed on a block.
+    By default, the hash of block MUST match the block id used for fetching it.
+    Optionally, an encryption/decryption operation could be implemented,
+    when fetching or eviting a block.
+*/
+static sebio_ctx_t sebio_ctx = {
+    {},
+    SEBIO_NO_CRYPTO,
+    &sebio_fetch_from_block_store,
+    &sebio_evict_to_block_store
+};
+//########################################################################
+
+/*
+    Set the context for the secure block IO
+*/
+state_status_t sebio_set(sebio_ctx_t ctx) {
+    sebio_ctx = ctx;
+    if(sebio_ctx.f_sebio_fetch == NULL || sebio_ctx.f_sebio_evict == NULL) {
+        sebio_ctx.f_sebio_fetch = &sebio_fetch_from_block_store;
+        sebio_ctx.f_sebio_evict = sebio_evict_to_block_store;
+    }
+    return STATE_SUCCESS;
+}
+
+state_status_t sebio_fetch(
+    uint8_t* block_id,
+    size_t block_id_size,
+    sebio_crypto_algo_e crypto_algo,
+    uint8_t** block,
+    size_t* block_size)
+{
+    return sebio_ctx.f_sebio_fetch(block_id, block_id_size, crypto_algo, block, block_size);
+}
+
+state_status_t sebio_evict(
+    uint8_t* block,
+    size_t block_size,
+    sebio_crypto_algo_e crypto_algo,
+    ByteArray& idOnEviction)
+{
+    return sebio_ctx.f_sebio_evict(block, block_size, crypto_algo, idOnEviction);
+}
+
+/*
+    The fetch function gets a block from the block store.
+    It requests first the size of a block,
+    then it allocates memory to contain it and loads it,
+    then it hashes the block and checks that it matches the id/hash given by the caller,
+    finally it decrypts the block if the caller specified that and set a context.
+
+    CONVENTION: the memory allocated for the block MUST be freed by the caller.
```

Comment:
The semantics of this function could really use reworking in my opinion. 
I think this was originally done because the block store API you had before took uint8_t* bufs and worked with those. 

However, all of the APIs here now work directly with ByteArrays. Making this function work with this prototype involves malloc'ing a buffer and copying stuff into it for essentially no reason. It would be better to have the caller make a ByteArray for the resulting block, then pass that array into this function and have the function do a resize() on it as needed. This will remove the need for malloc and a bunch of copying.

Reply:
Suggest we accept and fix later.